web/api/v1/api.go

```diff
@@ -1515,6 +1607,77 @@ func parseExcludeAlerts(r *http.Request) (bool, error) {
 	return excludeAlerts, nil
 }
 
+func parseListRulesPaginationRequest(r *http.Request) (*listRulesPaginationRequest, *parsePaginationError) {
+	var (
+		returnMaxAlert      = int32(-1)
+		returnMaxRuleGroups = int32(-1)
+		returnNextToken     = ""
+	)
+
+	if r.URL.Query().Get("maxAlerts") != "" {
+		maxAlert, err := strconv.ParseInt(r.URL.Query().Get("maxAlerts"), 10, 32)
+		if err != nil || maxAlert < 0 {
+			return nil, &parsePaginationError{
+				err:       fmt.Errorf("maxAlerts need to be a valid number and larger than 0: %w", err),
+				parameter: "maxAlerts",
+			}
+		}
+		returnMaxAlert = int32(maxAlert)
+	}
+
+	if r.URL.Query().Get("maxRuleGroups") != "" {
+		maxRuleGroups, err := strconv.ParseInt(r.URL.Query().Get("maxRuleGroups"), 10, 32)
+		if err != nil || maxRuleGroups < 0 {
+			return nil, &parsePaginationError{
+				err:       fmt.Errorf("maxRuleGroups need to be a valid number and larger than 0: %w", err),
+				parameter: "maxAlerts",
+			}
+		}
+		returnMaxRuleGroups = int32(maxRuleGroups)
+	}
+
+	if r.URL.Query().Get("nextToken") != "" {
+		returnNextToken = r.URL.Query().Get("nextToken")
+	}
+
+	if returnMaxRuleGroups >= 0 || returnMaxAlert >= 0 || returnNextToken != "" {
+		return &listRulesPaginationRequest{
+			MaxRuleGroups: returnMaxRuleGroups,
+			MaxAlerts:     returnMaxAlert,
+			NextToken:     returnNextToken,
+		}, nil
+	}
+
+	return nil, nil
+}
+
+func TruncateGroupInfos(groupInfos []*RuleGroup, maxRuleGroups int) ([]*RuleGroup, string) {
+	resultNumber := 0
+	var returnPaginationToken string
+	returnGroupDescs := make([]*RuleGroup, 0, len(groupInfos))
+	for _, groupInfo := range groupInfos {
+		// Add the rule group to the return slice if the maxRuleGroups is not hit
+		if maxRuleGroups < 0 || resultNumber < maxRuleGroups {
+			returnGroupDescs = append(returnGroupDescs, groupInfo)
+			resultNumber++
+			continue
+		}
+
+		// Return the next token if there is more aggregation group
+		if maxRuleGroups > 0 && resultNumber == maxRuleGroups {
+			returnPaginationToken = getRuleGroupNextToken(returnGroupDescs[maxRuleGroups-1].File, returnGroupDescs[maxRuleGroups-1].Name)
+			break
+		}
+	}
+	return returnGroupDescs, returnPaginationToken
+}
+
+func getRuleGroupNextToken(namespace, group string) string {
+	h := sha1.New()
```

Comment:
We will need to figure out how to handle this.

Reply:
we could use base64 encoding instead of hash, but this would also not keep the order that users created the rule groups in the files. I wonder if we really need to sort the output and instead we should rely on the position of the rule groups to paginate.

Reply:
Going back to this question then:

> Is there a good reason why we actually need the token to look like a hash value vs. just using the unhashed file + group name concatenation as a token that would preserve the original ordering?

I think we should definitely try to preserve ordering. Might not always matter, but I can imagine some users will have their groups in orders that have some meaning for them, and a larger subset of users will just find it confusing to not have the order preserved.

Reply:
Agreed. I tried to address that. I converted the algorithm to work with the position of the rule group.

Reply:
I assume this is resolved, I can see we simply hash file and group name as keys and then hash every group on LIST to find a matching one, right?

Reply:
that is correct